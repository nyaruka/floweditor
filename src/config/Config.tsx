import * as React from 'react';
import { ConfigProviderContext, assetHostPT, endpointsPT, languagesPT } from './ConfigProvider';

interface ConfigProps {
    render: (context: ConfigProviderContext) => React.ReactNode;
}

export default class Config extends React.Component<ConfigProps> {
    public static contextTypes = {
        assetHost: assetHostPT,
        endpoints: endpointsPT,
        languages: languagesPT
    };

    public render(): React.ReactNode {
        return this.props.render(this.context);
    }
}
