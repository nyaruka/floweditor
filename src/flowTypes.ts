import {
    IsOptionUniqueHandler,
    IsValidNewOptionHandler,
    NewOptionCreatorHandler,
    PromptTextCreatorHandler
} from 'react-select';

export interface Languages {
    [iso: string]: string;
}

export interface Endpoints {
    fields: string;
    groups: string;
    engine: string;
    contacts: string;
    flows: string;
    activity: string;
}

export interface FlowEditorConfig {
    languages: { [iso: string]: string };
    endpoints: Endpoints;
    flow: string;
    path?: string;
}

export interface LocalizationMap {
    [lang: string]: {
        [uuid: string]: any;
    };
}

export interface FlowDefinition {
    localization: LocalizationMap;
    language: string;
    name: string;
    nodes: FlowNode[];
    uuid: string;
    _ui: UIMetaData;
}

export interface FlowNode {
    uuid: string;
    actions: Action[];
    exits: Exit[];
    router?: Router;
    wait?: any;
}

export interface Exit {
    uuid: string;
    name?: string;
    destination_node_uuid?: string;
}

export interface Router {
    type: string;
    result_name?: string;
}

export interface Case {
    uuid: string;
    type: string;
    exit_uuid: string;
    arguments?: string[];
}

export interface SwitchRouter extends Router {
    cases: Case[];
    operand: string;
    default_exit_uuid: string;
}

export enum WaitType {
    exp = 'exp',
    group = 'group',
    msg = 'msg',
    flow = 'flow'
}

export interface Wait {
    type: WaitType;
    flow_uuid?: string;
}

export interface Group {
    uuid: string;
    name: string;
}

export enum Methods {
    GET = 'GET',
    POST = 'POST',
    PUT = 'PUT'
}

export interface Action {
    type: string;
    uuid: string;
}

export interface ChangeGroups extends Action {
    groups: Group[];
}

export interface Field {
    key: string;
    name: string;
}

export interface SetContactField extends Action {
    field: Field;
    value: string;
}

export interface SetContactProperty extends Action {
    property: string;
    value: string;
}

export type SetContactAttribute = SetContactField | SetContactProperty;

export interface SendMsg extends Action {
    text: string;
    all_urns?: boolean;
}

export interface SetPreferredChannel extends Action {
    language: string;
}

export interface SendEmail extends Action {
    subject: string;
    body: string;
    emails: string[];
}

export interface SetRunResult extends Action {
    result_name: string;
    value: string;
    category?: string;
}

export interface Headers {
    [name: string]: string;
}

export interface CallWebhook extends Action {
    url: string;
    method: Methods;
    body?: string;
    headers?: Headers;
}

export interface StartFlow extends Action {
    flow_name: string;
    flow_uuid: string;
}

export interface UIMetaData {
    nodes: { [key: string]: UINode };
    languages: Array<{ [iso: string]: string }>;
}

export interface FlowPosition {
    left: number;
    top: number;
    right?: number;
    bottom?: number;
}

export interface Dimensions {
    width: number;
    height: number;
}

export interface UINode {
    position: FlowPosition;

    // ui type, used for split by expression, contact field, etc
    type?: string;
}

export interface StickyNote {
    position: FlowPosition;
    title: string;
    body: string;
    color?: string;
}

export interface UIMetaData {
    nodes: { [key: string]: UINode };
    stickies: { [key: string]: StickyNote };
}

export type AnyAction =
    | Action
    | ChangeGroups
    | SetContactField
    | SetContactProperty
    | SetRunResult
    | SendMsg
    | SetPreferredChannel
    | SendEmail
    | CallWebhook
    | StartFlow;

export enum ContactProperties {
    UUID = 'UUID',
    'Created By' = 'Created By',
    'Modified By' = 'Modified By',
    Org = 'Org',
    Name = 'Name',
    Language = 'Language',
    Timezone = 'Timezone',
    Email = 'Email',
    Mailto = 'Mailto',
    Phone = 'Phone',
    Groups = 'Groups',
    Facebook = 'Facebook',
    Telegram = 'Telegram'
}

export enum ResultType {
    flow = 'flow',
    field = 'field',
    group = 'group',
    label = 'label'
}

export enum ValueType {
    text = 'text',
    numeric = 'numeric',
    datetime = 'datetime',
    state = 'state',
    district = 'district',
    ward = 'ward'
}

export enum AttributeType {
    property = 'property',
    field = 'field'
}

export interface CreateOptions {
    promptTextCreator?: PromptTextCreatorHandler;
    newOptionCreator?: NewOptionCreatorHandler;
    isValidNewOption?: IsValidNewOptionHandler;
    isOptionUnique?: IsOptionUniqueHandler;
    createNewOption?: NewOptionCreatorHandler;
    createPrompt?: string;
}
