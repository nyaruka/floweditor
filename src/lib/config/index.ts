import ConfigProvider, {
    ConfigProviderContext,
    endpointsPT,
    languagesPT
} from './ConfigProvider';
import {
    Type,
    Mode,
    ADD_GROUP_ACTION_TYPE,
    REMOVE_GROUP_ACTION_TYPE,
    typeConfigList,
    getTypeConfig
} from './typeConfigs';
import {
    Operator,
    operatorConfigList,
    getOperatorConfig
} from '../config/operatorConfigs';

export {
    ConfigProvider,
    ConfigProviderContext,
    Type,
    Operator,
    Mode,
    endpointsPT,
    languagesPT,
    ADD_GROUP_ACTION_TYPE,
    REMOVE_GROUP_ACTION_TYPE,
    typeConfigList,
    operatorConfigList,
    getTypeConfig,
    getOperatorConfig
};
