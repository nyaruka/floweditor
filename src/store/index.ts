import {
    UpdateCreateNodePosition,
    UpdateDragGroup,
    UpdateLanguage,
    UpdateNodeDragging,
    UpdateNodeEditorOpen,
    UpdateOperand,
    UpdateResultName,
    UpdateShowResultName,
    UpdateTranslating,
    UpdateTypeConfig,
    UpdateUserAddingAction,
    UpdateContactFieldsAction,
    UpdateContactFields,
    UpdateGroups,
    UpdateDragSelection
} from './actionTypes';
import createStore from './createStore';
import Constants from './constants';
import {
    CompletionOption,
    SearchResult,
    updateDefinition,
    updateDependencies,
    updateContactFields
} from './flowContext';
import {
    Flows,
    updateCreateNodePosition,
    updateDragGroup,
    updateFetchingFlow,
    updateFlows,
    updateGhostNode,
    updateLanguage,
    updateNodeDragging,
    updateNodeEditorOpen,
    updateTranslating,
    updateDragSelection
} from './flowEditor';
import {
    updateActionToEdit,
    updateOperand,
    updateResultName,
    updateShowResultName,
    updateTypeConfig,
    updateUserAddingAction
} from './nodeEditor';
import AppState, { initialState } from './state';
import {
    ActionAC,
    ConnectionEvent,
    DisconnectExit,
    disconnectExit,
    DispatchWithState,
    ensureStartNode,
    EnsureStartNode,
    fetchFlow,
    FetchFlow,
    fetchFlows,
    FetchFlows,
    GetState,
    LocalizationUpdates,
    moveActionUp,
    NoParamsAC,
    onAddToNode,
    OnAddToNode,
    onConnectionDrag,
    OnConnectionDrag,
    OnNodeMoved,
    onNodeMoved,
    OnOpenNodeEditor,
    onOpenNodeEditor,
    onUpdateAction,
    OnUpdateAction,
    OnUpdateLocalizations,
    onUpdateLocalizations,
    OnUpdateRouter,
    onUpdateRouter,
    removeAction,
    RemoveNode,
    removeNode,
    resetNodeEditingState,
    UpdateConnection,
    updateConnection,
    UpdateDimensions,
    updateDimensions,
    updateSticky,
    UpdateSticky
} from './thunks';

export {
    AppState,
    CompletionOption,
    SearchResult,
    Constants,
    DispatchWithState,
    GetState,
    ConnectionEvent,
    LocalizationUpdates,
    disconnectExit,
    onUpdateRouter,
    onUpdateAction,
    initialState,
    Flows,
    UpdateLanguage,
    UpdateTranslating,
    UpdateNodeEditorOpen,
    UpdateResultName,
    UpdateOperand,
    UpdateDragGroup,
    UpdateCreateNodePosition,
    createStore,
    updateTranslating,
    updateDragGroup,
    updateUserAddingAction,
    updateTypeConfig,
    UpdateUserAddingAction,
    UpdateContactFieldsAction,
    UpdateContactFields,
    UpdateGroups,
    UpdateTypeConfig,
    updateOperand,
    updateResultName,
    updateLanguage,
    updateFetchingFlow,
    updateDefinition,
    updateNodeDragging,
    updateFlows,
    updateDependencies,
    updateContactFields,
    updateNodeEditorOpen,
    updateGhostNode,
    updateCreateNodePosition,
    updateActionToEdit,
    fetchFlow,
    FetchFlow,
    FetchFlows,
    fetchFlows,
    ensureStartNode,
    EnsureStartNode,
    updateConnection,
    onUpdateLocalizations,
    OnUpdateLocalizations,
    DisconnectExit,
    ActionAC,
    UpdateConnection,
    OnConnectionDrag,
    resetNodeEditingState,
    onConnectionDrag,
    OnUpdateRouter,
    OnUpdateAction,
    onAddToNode,
    OnAddToNode,
    onNodeMoved,
    OnNodeMoved,
    onOpenNodeEditor,
    OnOpenNodeEditor,
    NoParamsAC,
    removeNode,
    RemoveNode,
    updateDimensions,
    UpdateDimensions,
    moveActionUp,
    removeAction,
    updateShowResultName,
    UpdateShowResultName,
    UpdateNodeDragging,
    UpdateDragSelection,
    updateDragSelection,
    updateSticky,
    UpdateSticky
};
