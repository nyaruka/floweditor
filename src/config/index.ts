import ConfigProvider, {
    ConfigProviderContext,
    assetHostPT,
    endpointsPT,
    languagesPT
} from './ConfigProvider';
import Config from './Config';
import { Type, Mode, typeConfigList, getTypeConfig } from './typeConfigs';
import { Operator, operatorConfigList, getOperatorConfig } from './operatorConfigs';

export {
    ConfigProviderContext,
    Type,
    Mode,
    Operator,
    assetHostPT,
    endpointsPT,
    languagesPT,
    typeConfigList,
    getTypeConfig,
    operatorConfigList,
    getOperatorConfig,
    Config
};

export default ConfigProvider;
