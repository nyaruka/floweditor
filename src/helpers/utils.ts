import { ComponentClass, SFC, ReactElement } from 'react';
import { ShallowWrapper, ReactWrapper } from 'enzyme';
import { FlowDefinition } from '../flowTypes';
import { CharacterSet } from '../components/form/TextInputElement';

const SNAKED_CHARS = /\s+(?=\S)/g;
const V4_UUID = /^[0-9A-F]{8}-[0-9A-F]{4}-4[0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}$/i;

interface BoolMap {
    [key: string]: boolean;
}

/**
 * Turns a string array into a bool map for constant lookup
 * @param {string[]} array - an array of strings, e.g. contact fields
 * @returns {object} A map of each string
 */
export const toBoolMap = (array: string[]): BoolMap =>
    array.reduce(
        (map: BoolMap, item: string) => ({
            ...map,
            [item]: true
        }),
        {}
    );

/**
 * Inserts commas into numbers where appropriate for better readability
 * @param {number} value - A number, e.g. 10000
 * @returns {string} A comma-separated string, e.g. 10,000
 */
export const addCommas = (value: number): string =>
    value.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ',');

/**
 * Replaces spaces with underscores
 * @param {string} value - A space-separated string to be snaked, e.g. a raw flow field name ('my flow field')
 * @returns {string} A snaked string, e.g. 'my_flow_field'
 */
export const snakify = (value: string): string =>
    value
        .toLowerCase()
        .trim()
        .replace(SNAKED_CHARS, '_');

/**
 * Returns a React component's name so we can attach it to a HOC's displayName property and view it in RDT
 * @param {ComponentClass | SFC} Component - A React component
 * @returns {string} The component's name
 */
export const getDisplayName = (WrappedComponent: ComponentClass | SFC): string =>
    WrappedComponent.displayName || WrappedComponent.name || 'Component';

/**
 * NOTE: borrowed from EventBrite: https://github.com/eventbrite/javascript/blob/master/react/testing.md#finding-nodes
 * Finds all instances of components in the rendered `componentWrapper` that are DOM components
 * with the `data-spec` attribute matching `name`.
 * @param {ReactWrapper|ShallowWrapper} componentWrapper - Rendered componentWrapper (result of mount, shallow, or render)
 * @param  {string} snacName - Name of `data-spec` attribute value to find
 * @param {string|Function} (Optional) typeFilter - (Optional) Expected type of the wrappers (defaults to all HTML tags)
 * @returns {ReactWrapper|ReactWrapper[]|ShallowWrapper|ShallowWrapper[]} Matching DOM components
 */
export const getSpecWrapper = (
    componentWrapper: ReactWrapper | ShallowWrapper,
    specName: string,
    attributeName: string = 'data-spec'
): any => {
    return componentWrapper.find(`[${attributeName}="${specName}"]`);
};

/**
 * Returns true if a given UUID matches v4 format
 * @param {string} uuid - A version 4 UUID (no braces, uppercase OK)
 * @returns {boolean}
 */
export const validUUID = (uuid: string): boolean => V4_UUID.test(uuid);

/**
 * Returns a given string in title case, e.g. 'full name' becomes 'Full Name'
 * @param {string} str - string to be title-cased
 * @returns {string} Title-cased string
 */
export const titleCase = (str: string): string =>
    str.replace(/\b\w+/g, s => s.charAt(0).toUpperCase() + s.substr(1).toLowerCase());

export const getSelectClass = (errors: number): string => {
    if (errors === 0) {
        return '';
    }
    // We use a global selector here for react-select
    return 'select-invalid';
};

export const reorderList = (list: any[], startIndex: number, endIndex: number): any[] => {
    const [removed] = list.splice(startIndex, 1);

    list.splice(endIndex, 0, removed);

    return list;
};

/**
 * Compares basic objects (no methods and DOM nodes; property order important)
 * @param {object} objA - basic object
 * @param {object} objB - basic object
 * @returns {boolean}
 */
export const jsonEqual = (objA: {}, objB: {}): boolean =>
    JSON.stringify(objA) === JSON.stringify(objB);

/**
 * Checks whether any of a list of error strings contain one or more queries.
 * Used to determine whether a particular error has been encountered.
 * @param {Array.<string>} errors - list of error messages
 * @param {Array.<string>} queries  - list of queries
 * @returns {boolean}
 */
export const hasErrorType = (errors: string[], queries: string[]): boolean => {
    if (!errors.length) {
        return false;
    }

    for (const error of errors) {
        for (const query of queries) {
            if (error.indexOf(query) > -1) {
                return true;
            }
        }
    }

    return false;
};
