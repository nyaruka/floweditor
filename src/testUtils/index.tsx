/* eslint-disable @typescript-eslint/no-explicit-any */
// TODO: Remove use of Function
// tslint:disable:ban-types
import { ConfigProviderContext, fakePropType } from 'config/ConfigProvider';
import { FlowTypes } from 'config/interfaces';
import { info } from 'core-js/core/log';
import { keyFor } from 'core-js/fn/symbol';
import { mount, ReactWrapper, shallow, ShallowWrapper } from 'enzyme';
import { FlowDefinition, FlowEditorConfig } from 'flowTypes';
import mutate, { Query } from 'immutability-helper';
import * as React from 'react';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { loadStore } from 'store';
import createStore from 'store/createStore';
import { RenderNodeMap } from 'store/flowContext';
import { getFlowComponents } from 'store/helpers';
import AppState, { initialState } from 'store/state';
import { App, FlowContents, FlowInfo, InfoResult, TembaAppState } from 'temba-components';
import { getFlowEditorConfig } from 'test/config';
import * as matchers from 'testUtils/matchers';
import { createUUID, merge, set } from 'utils';

// we need to use require syntax to bust implicit any
const boring: FlowDefinition = require('test/flows/boring.json');

// force our matchers to be read in
const match = matchers;

export interface Resp {
  assets: Array<{ [key: string]: any }>;
}

export interface QueryString {
  [key: string]: string;
}

export const contextTypes: { [key: string]: Function } = {
  config: fakePropType,
  store: fakePropType,
  assetService: fakePropType
};

export const baseState: AppState = mutate(initialState, {
  flowContext: merge({
    definition: require('test/flows/colors.json') as FlowDefinition
  }),
  editorState: merge({
    language: { id: 'eng', name: 'English' }
  })
});

export const getConfigProviderContext = (
  flowType: FlowTypes = FlowTypes.MESSAGING
): ConfigProviderContext => {
  const flowEditorConfig = getFlowEditorConfig(flowType);
  const win = window as any;
  if (win.isMobile && win.isMobile()) {
    flowEditorConfig.mutable = false;
  }

  const configProviderContext = { config: flowEditorConfig };
  return configProviderContext;
};

export const setMock = (implementation?: (...args: any[]) => any): Query<jest.Mock> =>
  set(jest.fn(implementation));

/**
 * Compose setup method for component tests
 */
export const composeSetup = <P extends {}>(
  Component: React.ComponentClass | React.SFC,
  baseProps: P = {} as any,
  flowType: FlowTypes = FlowTypes.MESSAGING,
  baseDuxState: AppState | Partial<AppState> = baseState,
  baseContext: ConfigProviderContext = getConfigProviderContext(flowType)
) => (
  shallowRender: boolean = true,
  propOverrides: Query<P | Partial<P>> = {},
  duxStateOverrides: Query<AppState | Partial<AppState>> = {},
  contextOverrides: Query<ConfigProviderContext | Partial<ConfigProviderContext>> = {},
  childContextTypeOverrides: { [key: string]: Function } = {},
  children: Array<JSX.Element | React.ComponentClass | React.SFC> = []
) => {
  const props = mutate(baseProps, propOverrides);
  const store = createStore(mutate(baseDuxState, duxStateOverrides) as AppState);
  let context = mutate(baseContext, merge({ store }));

  if (Object.keys(duxStateOverrides).length > 0) {
    context = mutate(context, duxStateOverrides);
  }

  if (Object.keys(contextOverrides).length > 0) {
    context = mutate(context, contextOverrides as ConfigProviderContext);
  }

  const childContextTypes: { [contextProp: string]: Function } = mutate(
    contextTypes,
    childContextTypeOverrides
  );
  // tslint:disable-next-line:ban-types
  const wrapper = (shallowRender ? (shallow as Function) : (mount as Function))(
    <Component {...props}>{children}</Component>,
    {
      context,
      childContextTypes
    }
  );

  return {
    wrapper,
    props,
    context,
    instance: wrapper.instance()
  };
};

export const composeSpy = (obj: Object | React.ComponentClass) => (instanceMethod: string) =>
  jest.spyOn((obj as React.ComponentClass).prototype || obj, instanceMethod as any);

/**
 * Wait for promises in queue to resolve
 */
export const flushPromises = () => new Promise((resolve: any) => setImmediate(resolve));

/**
 * Restore spy mocks (distinct from mocks created w/ jest.fn()).
 * Use to declaratively restore multiple spy mocks,
 * otherwise just call .mockRestore() on lone spy.
 */
export const restoreSpies = (...spies: Array<jest.SpyInstance<any>>) =>
  spies.forEach((spy: any) => spy.mockRestore());

// To-do: type this method's output, can pass it prop generic ingested by getComponentTestUtils
/**
 * NOTE: borrowed from EventBrite: https://github.com/eventbrite/javascript/blob/master/react/testing.md#finding-nodes
 * Finds all instances of components in the rendered `componentWrapper` that are DOM components
 * with the `data-spec` attribute matching `name`.
 * @param {ReactWrapper|ShallowWrapper} componentWrapper - Rendered componentWrapper (result of mount, shallow, or render)
 * @param  {string} snacName - Name of `data-spec` attribute value to find
 * @param {string|Function} (Optional) typeFilter - (Optional) Expected type of the wrappers (defaults to all HTML tags)
 * @returns {ReactWrapper|ReactWrapper[]|ShallowWrapper|ShallowWrapper[]} Matching DOM components
 */
export const getSpecWrapper = (
  componentWrapper: ReactWrapper<{}, {}> | ShallowWrapper<{}, {}>,
  specName: string,
  attributeName: string = 'data-spec'
): any => componentWrapper.find(`[${attributeName}="${specName}"]`);

export const composeDuxState = (
  query: Query<AppState | Partial<AppState>> = {},
  duxState = baseState
) => mutate(duxState, query);

export const composeComponentTestUtils = <P extends {}>(
  Component: React.ComponentClass | React.SFC | any,
  baseProps: P = {} as any,
  flowType: FlowTypes = FlowTypes.MESSAGING,
  baseDuxState: AppState | Partial<AppState> = baseState,
  baseContext: ConfigProviderContext = getConfigProviderContext(flowType)
) => ({
  setup: composeSetup<P>(Component, baseProps, flowType, baseDuxState, baseContext),
  spyOn: composeSpy(Component)
});

export const prepMockDuxState = (): {
  testNodes: RenderNodeMap;
  mockDuxState: AppState;
} => {
  const testNodes = getFlowComponents(boring).renderNodeMap;

  return {
    testNodes,
    mockDuxState: {
      ...initialState,
      flowContext: {
        ...initialState.flowContext,
        definition: boring,
        nodes: testNodes
      }
    }
  };
};

export const createMockStore: Function = configureStore([thunk]);

export const mock = <T extends {}, K extends keyof T>(object: T, property: K, value: T[K]) => {
  Object.defineProperty(object, property, { get: () => value });
};

interface StoreOverrides {
  languageCode?: string;
  isTranslating?: boolean;
  results?: InfoResult[];
}

export const setupStore = ({
  languageCode = 'eng',
  isTranslating = false,
  results = []
}: StoreOverrides = {}) => {
  const languageNames: { [key: string]: string } = { eng: 'English', spa: 'Spanish' };
  const store = loadStore();
  mock(store, 'getState', () => {
    const state: TembaAppState = {
      isTranslating,
      languageCode,
      languageNames,
      workspace: {
        uuid: createUUID(),
        name: 'Test Workspace',
        languages: ['eng', 'spa'],
        timezone: 'UTC',
        date_style: 'MM/DD/YYYY',
        country: 'US',
        anon: false
      },
      flowDefinition: null,
      flowInfo: null,
      
      // Editor state from Redux
      editorState: null,
      
      // Flow context from Redux  
      flowNodes: {},
      flowMetadata: null,
      assetStore: null,
      flowIssues: {},
      
      // Node editor state from Redux
      nodeEditorSettings: null,
      nodeEditorTypeConfig: null,
      userAddingAction: false,
      
      setFlowInfo: (info: FlowInfo) => {},
      setFlowContents: (contents: FlowContents) => {},
      setLanguageCode: (code: string) => {},
      getLanguage: () => ({ name: languageNames[languageCode], code: languageCode }),
      getFlowResults: () => results,
      getResultByKey: (key: any) => {
        const result = results.find(r => r.key === key);
        return result || null;
      },
      updateCanvasPositions: (positions: any) => {},
      removeNodes: (uuids: string[]) => {},
      
      // Editor state management
      updateEditorState: (changes: any) => {},
      
      // Flow context management
      updateFlowDefinition: (definition: any) => {},
      updateFlowNodes: (nodes: any) => {},
      updateAssetStore: (assetStore: any) => {},
      updateFlowIssues: (issues: any) => {},
      updateFlowMetadata: (metadata: any) => {},
      
      // Node editor management
      updateNodeEditorSettings: (settings: any) => {},
      updateNodeEditorTypeConfig: (typeConfig: any) => {},
      updateUserAddingAction: (userAddingAction: boolean) => {}
    };
    return state;
  });

  mock(store, 'getApp', () => {
    const app: App = {
      subscribe: (callback: (state: any, prevState: any) => void) => {
        // Mock implementation of subscribe
        return () => {
          // Mock unsubscribe function
        };
      }
    };
    return app;
  });

  mock(store, 'resolveUsers', async (objects: any[], fields: string[]) => {
    // mock implementation to add UUIDs to users that don't have them
    objects.forEach(obj => {
      fields.forEach(field => {
        if (obj[field] && obj[field].email && !obj[field].uuid) {
          obj[field].uuid = `mock-uuid-${obj[field].email.replace('@', '-at-')}`;
        }
      });
    });
  });
};
