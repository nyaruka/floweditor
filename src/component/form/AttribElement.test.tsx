import * as config from '../../../assets/config';
import { AttributeType, ResultType } from '../../flowTypes';
import { SearchResult } from '../../store';
import { createSetup, createSpy } from '../../testUtils';
import { getSelectClass } from '../../utils';
import { V4_UUID } from '../../utils';
import {
    AttribElement,
    AttribElementProps,
    CREATE_PROMPT,
    attribExists,
    fieldNameValid,
    NOT_FOUND,
    PLACEHOLDER,
    isValidNewOption,
    isOptionUnique,
    createNewOption
} from './AttribElement';

const initial: SearchResult = {
    id: 'name',
    name: 'Name',
    type: AttributeType.property
};

const baseProps: AttribElementProps = {
    name: 'Attribute',
    contactFields: [],
    initial,
    endpoint: config.endpoints.fields
};

const setup = createSetup<AttribElementProps>(AttribElement, baseProps);

const spyOn = createSpy(AttribElement);

describe(`${AttribElement.name}`, () => {
    describe('helpers', () => {
        describe('attribExists', () => {
            const matchingOptions = [
                {
                    name: 'Expected Delivery Date',
                    id: 'expected_delivery_date',
                    type: 'field'
                }
            ];

            it('should return true if field exists in matching options provided by react-select', () => {
                expect(attribExists('expected delivery date', matchingOptions)).toBeTruthy();
            });

            it('should return false if field does not exist in matching options provided by react-select', () => {
                expect(attribExists('national id', [])).toBeFalsy();
                expect(attribExists('national id', matchingOptions)).toBeFalsy();
            });
        });

        describe('fieldNameValid', () => {
            it('should return true if field name is valid', () => {
                expect(fieldNameValid('Age')).toBeTruthy();
            });

            it('should return false if field name invalid', () => {
                expect(fieldNameValid('')).toBeFalsy();
                expect(
                    fieldNameValid('pneumonoultramicroscopicsilicovolcanoconiosis ')
                ).toBeFalsy();
                expect(fieldNameValid('Age$')).toBeFalsy();
            });
        });

        describe('isOptionUnique', () => {
            const isOptionUniqueSignature = {
                labelKey: 'name',
                valueKey: 'id',
                options: []
            };

            it('should return true if new option is unique', () => {
                const newOption = {
                    id: '2e020526-06a7-4acc-8f3f-90b4ceffdd91',
                    name: 'Age',
                    type: AttributeType.field
                };

                expect(
                    isOptionUnique({
                        ...isOptionUniqueSignature,
                        option: newOption
                    })
                ).toBeTruthy();
            });

            it('should return false if new option is not unique', () => {
                const newOption = {
                    id: 'name',
                    name: 'Name',
                    type: AttributeType.property
                };

                expect(
                    isOptionUnique({
                        ...isOptionUniqueSignature,
                        option: newOption
                    })
                ).toBeFalsy();
            });
        });

        describe('createNewOption', () => {
            it('should return a new SearchResult', () => {
                const newOption = { label: 'Age', labelKey: 'name', valueKey: 'id' };

                expect(createNewOption(newOption)).toEqual({
                    id: expect.stringMatching(V4_UUID),
                    name: newOption.label,
                    type: AttributeType.field,
                    extraResult: true
                });
            });
        });
    });

    describe('render', () => {
        it('should render self, children with base props', () => {
            const {
                wrapper,
                instance,
                props: { showLabel, name, helpText, endpoint, contactFields }
            } = setup({}, true);

            expect(wrapper.find('FormElement').props()).toEqual(
                expect.objectContaining({
                    showLabel,
                    name,
                    helpText,
                    errors: [],
                    attribError: false
                })
            );
            expect(wrapper.find('SelectSearch').props()).toEqual(
                expect.objectContaining({
                    __className: getSelectClass(0),
                    onChange: instance.onChange,
                    name,
                    url: endpoint,
                    resultType: ResultType.field,
                    localSearchOptions: contactFields,
                    multi: false,
                    initial: [initial],
                    closeOnSelect: true,
                    searchPromptText: NOT_FOUND,
                    placeholder: PLACEHOLDER
                })
            );
            expect(wrapper).toMatchSnapshot();
        });

        it('should pass createOptions to SelectSearch', () => {
            const { wrapper } = setup({ add: true }, true);

            expect(wrapper.find('SelectSearch').props()).toEqual(
                expect.objectContaining({
                    isValidNewOption,
                    isOptionUnique,
                    createNewOption,
                    createPrompt: CREATE_PROMPT
                })
            );
            expect(wrapper).toMatchSnapshot();
        });
    });

    describe('instance methods', () => {
        const existingField: SearchResult = {
            id: '2003ec76-69e3-455e-a603-938ad90cb53f',
            name: 'National ID',
            type: AttributeType.field
        };

        describe('onChange', () => {
            it('should set state if attribute is new', () => {
                const setStateSpy = spyOn('setState');
                const { wrapper, instance } = setup({}, true);

                instance.onChange(existingField);

                expect(setStateSpy).toHaveBeenCalledTimes(1);
                expect(setStateSpy).toHaveBeenCalledWith({ attribute: existingField });

                setStateSpy.mockRestore();
            });

            it('should not set state if attribute is not new', () => {
                const setStateSpy = spyOn('setState');
                // tslint:disable-next-line:no-shadowed-variable
                const { wrapper, instance, props: { initial } } = setup({}, true);

                instance.onChange(initial);

                expect(setStateSpy).toHaveBeenCalledTimes(0);

                setStateSpy.mockRestore();
            });
        });

        describe('getErrors', () => {
            it('should return list of errors', () => {
                const { wrapper, instance, props: { name } } = setup(
                    { required: true, initial: { ...initial, name: '' } },
                    true
                );

                expect(instance.getErrors()).toEqual([`${name} is required.`]);
            });

            it('should return an empty list', () => {
                const { wrapper, instance } = setup({}, true);

                expect(instance.getErrors()).toEqual([]);
            });
        });

        describe('updateErrorState', () => {
            it('should set state', () => {
                const setStateSpy = spyOn('setState');
                const { wrapper, instance } = setup({}, true);
                const oldErrorState = [];
                const newErrorState = [`${name} is required.`];

                instance.updateErrorState(newErrorState);

                expect(setStateSpy).toHaveBeenCalledTimes(1);
                expect(setStateSpy).toHaveBeenCalledWith({ errors: newErrorState });

                setStateSpy.mockRestore();
            });
        });

        describe('validate', () => {
            it('should return true if control does not contain errors', () => {
                const updateErrorStateSpy = spyOn('updateErrorState');
                const { wrapper, instance } = setup({}, true);

                expect(instance.validate()).toBeTruthy();
                expect(updateErrorStateSpy).toHaveBeenCalledTimes(1);
                expect(updateErrorStateSpy).toHaveBeenCalledWith([]);

                updateErrorStateSpy.mockRestore();
            });

            it('should return false if control contains errors', () => {
                const updateErrorStateSpy = spyOn('updateErrorState');
                const { wrapper, instance, props: { name } } = setup(
                    {
                        initial: { ...initial, name: '' },
                        required: true
                    },
                    true
                );

                expect(instance.validate()).toBeFalsy();
                expect(updateErrorStateSpy).toHaveBeenCalledTimes(1);
                expect(updateErrorStateSpy).toHaveBeenCalledWith([`${name} is required.`]);

                updateErrorStateSpy.mockRestore();
            });
        });
    });
});
