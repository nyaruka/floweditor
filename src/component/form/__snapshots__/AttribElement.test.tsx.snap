// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AttribElement render should pass createOptions to SelectSearch 1`] = `
<FormElement
  attribError={false}
  errors={Array []}
  name="Attribute"
>
  <SelectSearch
    __className=""
    closeOnSelect={true}
    createNewOption={[Function]}
    createPrompt="New attribute: "
    initial={
      Array [
        Object {
          "id": "name",
          "name": "Name",
          "type": "property",
        },
      ]
    }
    isOptionUnique={[Function]}
    isValidNewOption={[Function]}
    localSearchOptions={Array []}
    multi={false}
    name="Attribute"
    onChange={[Function]}
    placeholder="Enter the name of an existing attribute or create a new one"
    resultType="field"
    searchPromptText="Invalid attribute name"
    url="/assets/fields.json"
  />
</FormElement>
`;

exports[`AttribElement render should render self, children with base props 1`] = `
<FormElement
  attribError={false}
  errors={Array []}
  name="Attribute"
>
  <SelectSearch
    __className=""
    closeOnSelect={true}
    initial={
      Array [
        Object {
          "id": "name",
          "name": "Name",
          "type": "property",
        },
      ]
    }
    localSearchOptions={Array []}
    multi={false}
    name="Attribute"
    onChange={[Function]}
    placeholder="Enter the name of an existing attribute or create a new one"
    resultType="field"
    searchPromptText="Invalid attribute name"
    url="/assets/fields.json"
  />
</FormElement>
`;
